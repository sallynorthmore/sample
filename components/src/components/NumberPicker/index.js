import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import clamp from 'lodash/clamp';
import FilterControls from '../FilterControls';
import { IconPlus, IconMinus } from '../Icon';
import {
	Button,
	ButtonGroup,
	Inner,
	Input,
	InputGroup,
	Title,
	Unit,
	Validation,
} from './styles';

class NumberPicker extends Component {
	constructor(props) {
		super(props);
		this.state = {
			value: null,
		};
		this.debouncedOnValueChange = debounce(this.props.onValueChange, 200);
	}

	componentWillReceiveProps(nextProps) {
		if (this.state.value !== nextProps.value) {
			this.setState({
				value: nextProps.value,
			});
		}
	}

	getClampedValue(value) {
		const { minimum, maximum } = this.props;
		return clamp(value, minimum, maximum);
	}

	handleClearFilter = () => {
		this.handleValueChange(null);
	};

	handleDecrease = () => {
		this.handleValueDelta(-1);
	};

	handleIncrease = () => {
		this.handleValueDelta(1);
	};

	handleInputTextChange = event => {
		const stringValue = event.currentTarget.value;
		const numericValue = stringValue === '' ? null : parseFloat(stringValue);
		this.handleValueChange(numericValue);
	};

	handleValueChange(newValue) {
		if (this.isValueValid(newValue)) {
			this.debouncedOnValueChange({
				value: newValue === null ? null : newValue.toString(),
			});
		}

		this.setState({
			value: newValue,
		});
	}

	handleValueDelta(direction) {
		const { delta } = this.props;
		const { value } = this.state;
		this.handleValueChange(this.getClampedValue(value + delta * direction));
	}

	isValueValid(value) {
		return value === null || value === this.getClampedValue(value);
	}

	render() {
		const { title, minimum, maximum, unit } = this.props;
		const { value } = this.state;
		const isValid = this.isValueValid(value);

		return (
			<Inner>
				<FilterControls
					onClearClick={this.handleClearFilter}
					isClearDisabled={value === ''}
				/>

				{title && <Title>{title}</Title>}
				<Validation isVisible={!isValid}>
					({minimum} - {maximum})
				</Validation>
				<InputGroup>
					<Input
						name="input"
						onChange={this.handleInputTextChange}
						type="number"
						value={value ? value : ''}
					/>
					{unit && <Unit>{unit}</Unit>}
					<ButtonGroup>
						<Button name="increase" onClick={this.handleIncrease}>
							<IconPlus width={10} />
						</Button>
						<Button name="decrease" onClick={this.handleDecrease}>
							<IconMinus width={9} />
						</Button>
					</ButtonGroup>
				</InputGroup>
			</Inner>
		);
	}
}

NumberPicker.propTypes = {
	onValueChange: PropTypes.func.isRequired,
	title: PropTypes.string,
	value: PropTypes.number,
	delta: PropTypes.number.isRequired,
	minimum: PropTypes.number.isRequired,
	maximum: PropTypes.number.isRequired,
	unit: PropTypes.string,
};

NumberPicker.defaultProps = {
	delta: 1,
	minimum: -1000000000,
	maximum: 1000000000,
};

export default NumberPicker;
