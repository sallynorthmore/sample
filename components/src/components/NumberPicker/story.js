// import React from 'react';
// import { storiesOf } from '@storybook/react';
// import { withInfo } from '@storybook/addon-info';
// import { action } from '@storybook/addon-actions';
//
// import NumberPicker from './index';
//
// const stories = storiesOf('NumberPicker', module);
//
// stories
// 	.add(
// 		'Default',
// 		withInfo('NumberPicker documentation')(() => {
// 			return <NumberPicker onValueChange={action('Value changed')} />;
// 		})
// 	)
// 	.add(
// 		'with Title',
// 		withInfo('NumberPicker documentation')(() => {
// 			return (
// 				<NumberPicker
// 					title="Select your minimum size"
// 					onValueChange={action('Value changed')}
// 				/>
// 			);
// 		})
// 	)
// 	.add(
// 		'with inital value',
// 		withInfo('NumberPicker documentation')(() => {
// 			return (
// 				<NumberPicker
// 					title="Select your minimum size"
// 					value={100}
// 					onValueChange={action('Value changed')}
// 				/>
// 			);
// 		})
// 	)
// 	.add(
// 		'with unit set',
// 		withInfo('NumberPicker documentation')(() => {
// 			return (
// 				<NumberPicker
// 					title="Select your minimum size"
// 					value={100}
// 					unit="M"
// 					onValueChange={action('Value changed')}
// 				/>
// 			);
// 		})
// 	)
// 	.add(
// 		'with inital value, and delta set to 10',
// 		withInfo('NumberPicker documentation')(() => {
// 			return (
// 				<NumberPicker
// 					title="Select a number (in increments of 10)"
// 					value={100}
// 					delta={10}
// 					onValueChange={action('Value changed')}
// 				/>
// 			);
// 		})
// 	)
// 	.add(
// 		'with min and max values set',
// 		withInfo('NumberPicker documentation')(() => {
// 			return (
// 				<NumberPicker
// 					title="Select between 0 and 1,000"
// 					value={500}
// 					delta={100}
// 					minimum={0}
// 					maximum={1000}
// 					onValueChange={action('Value changed')}
// 				/>
// 			);
// 		})
// 	);
