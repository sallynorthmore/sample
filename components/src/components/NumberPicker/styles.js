import styled from 'styled-components';

export const Inner = styled.div`
	flex: 1;
	max-width: 400px;
	text-align: center;
`;

export const FilterButtonGroup = styled.div`
	box-shadow: inset 0 -1px 0 0 rgba(227, 232, 236, 1);
	height: 45px;
	text-align: right;
`;

export const Title = styled.p`
	color: rgba(122, 122, 122, 1);
	font: 400 13px 'Lato';
	letter-spacing: 0.63px;
	margin: 16px 0 0;
`;

export const Validation = styled.p`
	color: rgba(122, 122, 122, 1);
	font: 400 11px 'Lato';
	margin: 4px 0 0;
	visibility: ${props => (props.isVisible ? 'visible' : 'hidden')};
`;

export const InputGroup = styled.div`
	margin: 24px auto;
	padding: 8px 0;
	position: relative;
	width: 128px;
`;

export const Input = styled.input`
	background: white;
	border: none;
	box-shadow: inset 0 0 0 1px rgba(213, 216, 220, 1);
	color: rgba(122, 122, 122, 1);
	height: 40px;
	outline: none;
	padding: 0 15px 0 8px;
	transition: box-shadow 0.195s cubic-bezier(0.4, 0, 0.2, 1),
		color 0.195s cubic-bezier(0.4, 0, 0.2, 1);
	width: 100%;

	&:focus {
		box-shadow: inset 0 0 0 1px rgba(0, 129, 187, 1);
		color: rgba(0, 129, 187, 1);
	}

	&::selection {
		background: rgba(0, 129, 187, 1);
		color: white;
	}
`;

export const Unit = styled.div`
	background: rgba(255, 255, 255, 0.4);
	color: rgba(122, 122, 122, 1);
	display: block;
	font-size: 12px;
	font-weight: 400;
	padding: 4px 8px;
	position: absolute;
	right: 30px;
	top: 9px;
`;

export const ButtonGroup = styled.div`
	position: absolute;
	right: 0;
	top: 0;
	width: 30px;
`;

export const Button = styled.button`
	align-items: center;
	background: rgba(0, 129, 187, 1);
	border: none;
	color: white;
	cursor: pointer;
	display: flex;
	font: 700 16px/1 'Lato';
	height: 27px;
	justify-content: center;
	outline: none;
	position: relative;
	transition: background 0.192s cubic-bezier(0.4, 0, 0.2, 1);
	width: 30px;

	&:not(:first-child)::before {
		background: rgba(255, 255, 255, 0.2);
		content: '';
		display: block;
		height: 1px;
		left: 20%;
		position: absolute;
		top: 0;
		width: 60%;
	}

	&:hover {
		background: rgba(1, 167, 226, 1);

		/* stylelint-disable selector-max-type, selector-max-specificity */
		&:not(:first-child)::before,
		& + button:before {
			display: none;
		}
		/* stylelint-enable */
	}

	&:disabled {
		pointer-events: none;
	}
`;
