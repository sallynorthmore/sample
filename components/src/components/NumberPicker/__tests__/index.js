import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import NumberPicker from '../index';

const testProps = {
	delta: 10,
	maximum: 200,
	minimum: 0,
	onValueChange: () => null,
	value: 100,
};

describe('<NumberPicker /> snapshots', () => {
	it('renders correctly', () => {
		const tree = renderer.create(<NumberPicker {...testProps} />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
