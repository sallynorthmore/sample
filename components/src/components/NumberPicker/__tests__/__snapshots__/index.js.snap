// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberPicker /> snapshots renders correctly 1`] = `
.c1 {
  box-shadow: inset 0 -1px 0 0 rgba(227,232,236,1);
  -webkit-flex: 0 0 45px;
  -ms-flex: 0 0 45px;
  flex: 0 0 45px;
  text-align: right;
}

.c2 {
  background: transparent;
  border: none;
  color: rgba(122,122,122,0.7);
  cursor: pointer;
  display: inline-block;
  font: 400 13px/1 'Lato';
  opacity: 1;
  outline: none;
  padding: 16px 8px;
  pointer-events: auto;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2:hover {
  color: rgba(0,129,187,1);
}

.c2:not(:first-of-type)::before {
  background: rgba(227,232,236,1);
  content: '';
  display: inline-block;
  height: 13px;
  left: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 400px;
  text-align: center;
}

.c3 {
  color: rgba(122,122,122,1);
  font: 400 11px 'Lato';
  margin: 4px 0 0;
  visibility: hidden;
}

.c4 {
  margin: 24px auto;
  padding: 8px 0;
  position: relative;
  width: 128px;
}

.c5 {
  background: white;
  border: none;
  box-shadow: inset 0 0 0 1px rgba(213,216,220,1);
  color: rgba(122,122,122,1);
  height: 40px;
  outline: none;
  padding: 0 15px 0 8px;
  -webkit-transition: box-shadow 0.195s cubic-bezier(0.4,0,0.2,1), color 0.195s cubic-bezier(0.4,0,0.2,1);
  transition: box-shadow 0.195s cubic-bezier(0.4,0,0.2,1), color 0.195s cubic-bezier(0.4,0,0.2,1);
  width: 100%;
}

.c5:focus {
  box-shadow: inset 0 0 0 1px rgba(0,129,187,1);
  color: rgba(0,129,187,1);
}

.c5::selection {
  background: rgba(0,129,187,1);
  color: white;
}

.c6 {
  position: absolute;
  right: 0;
  top: 0;
  width: 30px;
}

.c7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgba(0,129,187,1);
  border: none;
  color: white;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font: 700 16px/1 'Lato';
  height: 27px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  position: relative;
  -webkit-transition: background 0.192s cubic-bezier(0.4,0,0.2,1);
  transition: background 0.192s cubic-bezier(0.4,0,0.2,1);
  width: 30px;
}

.c7:not(:first-child)::before {
  background: rgba(255,255,255,0.2);
  content: '';
  display: block;
  height: 1px;
  left: 20%;
  position: absolute;
  top: 0;
  width: 60%;
}

.c7:hover {
  background: rgba(1,167,226,1);
}

.c7:hover:not(:first-child)::before,
.c7:hover + button:before {
  display: none;
}

.c7:disabled {
  pointer-events: none;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <button
      className="c2"
      disabled={false}
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
  <p
    className="c3"
  >
    (
    0
     - 
    200
    )
  </p>
  <div
    className="c4"
  >
    <input
      className="c5"
      name="input"
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="c6"
    >
      <button
        className="c7"
        name="increase"
        onClick={[Function]}
      >
        <svg
          viewBox="0 0 10 10"
          width="10px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-103.000000, -12.000000)"
            >
              <g
                transform="translate(103.000000, 12.000000)"
              >
                <rect
                  height="10"
                  width="2"
                  x="4"
                  y="0"
                />
                <rect
                  height="10"
                  transform="translate(5.000000, 5.000000) rotate(90.000000) translate(-5.000000, -5.000000) "
                  width="2"
                  x="4"
                  y="0"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
      <button
        className="c7"
        name="decrease"
        onClick={[Function]}
      >
        <svg
          viewBox="0 0 9 2"
          width="9px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-104.000000, -42.000000)"
            >
              <rect
                height="9"
                transform="translate(108.500000, 43.000000) rotate(90.000000) translate(-108.500000, -43.000000) "
                width="2"
                x="107.5"
                y="38.5"
              />
            </g>
          </g>
        </svg>
      </button>
    </div>
  </div>
</div>
`;
